src/microsim/schema/simulation.py

```diff
@@ -90,6 +92,12 @@ def optical_image(
             raise ValueError("truth must be a DataArray")
         # let the given modality render the as an image (convolved, etc..)
         channel = self.channels[channel_idx]  # TODO
+        # TODO: Multi-fluorophore setup: do stochastic sampling on truth and pass the sampled output to self.modality.render.
+        # To achieve this, for every fluorophore, following needs to be done in excitation_emission_model.py:
+        # 1. Divide its excistion spectra into a avelength intervals. (Pre-computed for each fluorophore)
+        # 2. Sample the excitation of fluorophores within each interval on the basis of the incident light spectrum and its excitation spectra.
+        # TODO: figure out the math behind this.
+        # 3. Allocate the emitted light in waavelength intervals. (Pre-computed for each fluorophore)
```

Comment:
yep, totally agree on the steps here.

first, let's define what we're doing for a single pixel.  the pixel value coming in is number of fluorophores (per pixel), and we want the value going out to be the number of photons emitted per second (per pixel).  This will depend on the irradiance (W/cm^2) of the incoming light, and the ability of the fluorophore to absorb it (extinction coefficient), each of each will be a wavelength dependent property.

In other words, our incoming `FZYX` array axis gets converted to a `CWZYX` array, where `len(F) == number of fluorophores`, `len(W) == number of spectral bins`, and `len(C) == number of optical configurations` or "channels" ... (or, technically, C will be 1 at this point because the optical image function takes in a channel_index parameter and calculates it for one channel at a time)

on a high level, the number of photons emitted is going to be the [irrandiance](https://en.wikipedia.org/wiki/Irradiance) * the [fluorophore absorption cross section](https://en.wikipedia.org/wiki/Absorption_cross_section)  / [photon energy](https://en.wikipedia.org/wiki/Photon_energy)

Keeping all those units straight can be a big pain, so let's depend on `pint`.  I've done all these calculations a few times before, so I'll just paste what I've got below.  It would be great if you could double check it for me :joy: ... and see if it makes sense:

```python
from typing import Any

import numpy as np
import pint
from scipy.constants import Avogadro, c, h

from microsim.fpbase import get_fluorophore, get_microscope

ureg = pint.UnitRegistry()
AVOGADRO = Avogadro / ureg.mol
PLANCK = h * ureg.joule * ureg.second
C = c * ureg.meter / ureg.second


def get_overlapping_spectra(
    spectra: np.ndarray, spectrum2: np.ndarray
) -> tuple[np.ndarray, np.ndarray]:
    """Return overlapping subset of spectra.

    This assumes that the spectra are (N, 2) arrays where the first column is the
    wavelength and the second column is the intensity, and that the spectra are
    sorted by wavelength. (This is the format of the data returned by FPbase.)
    """
    # Find the indices of the start and end of the overlapping subset
    start = max(spectra[0, 0], spectrum2[0, 0])
    end = min(spectra[-1, 0], spectrum2[-1, 0])

    # Find the indices of the start and end of the overlapping subset
    start_idx = np.searchsorted(spectra[:, 0], start)
    end_idx = np.searchsorted(spectra[:, 0], end, side="right")

    start_idx2 = np.searchsorted(spectrum2[:, 0], start)
    end_idx2 = np.searchsorted(spectrum2[:, 0], end, side="right")
    return spectra[start_idx:end_idx], spectrum2[start_idx2:end_idx2]


def _ensure_quantity(value: Any, units: str) -> pint.Quantity:
    """Helper function to ensure that a value is a pint Quantity with `units`."""
    quant = ureg.Quantity(value)
    if quant.dimensionless:
        quant *= ureg.Quantity(units)
    _u = ureg.Quantity(units).units
    if not quant.is_compatible_with(_u):
        raise pint.DimensionalityError(quant.units, _u)
    return quant


def ec_to_cross_section(ec: Any) -> pint.Quantity:
    """Gives cross section in cm^2 from extinction coefficient in M^-1 * cm^-1."""
    ec = _ensure_quantity(ec, "cm^2/mol")
    # x1000?
    # this came from calculations elsewhere, and looking at wikipedia
    # and looking at Nathan Shaner's code
    # need to double check whether it's still correct with our units
    ec = ec * 1000
    return (ec * np.log(10) / AVOGADRO).to("cm^2")


def energy_per_photon(wavelength: Any) -> pint.Quantity:
    """Converts a wavelength to energy per photon in J."""
    wavelength = _ensure_quantity(wavelength, "1nm")
    return (PLANCK * C / wavelength).to("J")


def fluorophore_photon_flux(
    wavelength: Any,
    irradiance: Any,
    extinction_coefficient: Any,
) -> pint.Quantity:
    """Calculates the number of photons hitting a fluorophore per second."""
    irradiance = _ensure_quantity(irradiance, "1W/cm^2")
    E_photon = energy_per_photon(wavelength)
    cross_section = ec_to_cross_section(extinction_coefficient)
    exc_rate = (cross_section * irradiance / E_photon).to("1/s")

    return exc_rate


def get_emission_events(
    microscope_id: str,
    optical_config: str,
    fluorophore: str,
    *,
    light_power: float = 100,
) -> np.ndarray:
    """Calculate the number of emission events per fluorophore per second.

    This function takes an FPbase microscope ID, the name of an optical configuration,
    and the name of a fluorophore, and returns the number of emission events per second
    per fluorophore.
    """
    # get the microscope and fluorophore
    microscope = get_microscope(microscope_id)

    # find the excitation filter in the microscope
    try:
        WF_Green = next(
            oc for oc in microscope.opticalConfigs if oc.name == optical_config
        )
    except StopIteration as e:
        raise ValueError(
            f"Optical config {optical_config} not found in microscope {microscope_id}"
        ) from e

    try:
        ex_filter = next(f for f in WF_Green.filters if f.spectrum.subtype == "BP")
    except StopIteration as e:
        raise ValueError(
            f"Bandpass filter not found in optical config {optical_config}"
        ) from e

    # get the fluorophore
    fluor = get_fluorophore(fluorophore).default_state
    if not fluor:
        raise ValueError(f"Fluorophore {fluorophore} has no default state.")
    if not fluor.excitation_spectrum:
        raise ValueError(f"Fluorophore {fluorophore} has no excitation spectrum.")

    # convert the spectra to numpy arrays
    fluor_ex_spectrum = np.asarray(fluor.excitation_spectrum.data)
    filter_spectrum = np.asarray(ex_filter.spectrum.data)

    # find the subset of the spectra with overlapping wavelengths
    ex_spectrum, filter_spectrum = get_overlapping_spectra(
        fluor_ex_spectrum, filter_spectrum
    )

    # wavelengths in nm
    wavelengths = filter_spectrum[:, 0] * ureg.nm
    # power in units of W x cm-2 (irradiance), we're making up constant 100 for now...
    irradiance = filter_spectrum[:, 1] * _ensure_quantity(light_power, "W/cm^2")
    # convert ex_spectrum to extinction coefficient in units of M^-1 x cm^-1
    ext_coeff = ex_spectrum[:, 1] * fluor.extCoeff / ureg.molar / ureg.cm

    # calculate the number of photons hitting a fluorophore per second
    exc_rate = fluorophore_photon_flux(wavelengths, irradiance, ext_coeff)

    # if the fluorophore has a lifetime, calculate the effective excitation rate
    # based on the fraction of time in the excited state
    # (i.e. at some point, we'll hit ground state depletion and saturation effects)
    if (lifetime := getattr(fluor, "lifetime", None)) is not None:
        lifetime = _ensure_quantity(lifetime, "ns")
        # Calculate the fraction of time in the excited state
        f_excited = exc_rate * lifetime
        # Calculate the fraction of time in the ground state
        f_ground = 1 - f_excited
        # Calculate the effective excitation rate
        exc_rate = exc_rate * f_ground

    # multiply by the quantum yield to get the number of emission events
    emission_events = exc_rate * fluor.qy

    # recombine with wavelength:
    return wavelengths, emission_events


# we'll be calculating this page:
# https://www.fpbase.org/microscope/wKqWb/?c=Widefield%20Green&p=egfp_default
wavelengths, emission_events_per_fluor = get_emission_events(
    "wKqWb", "Widefield Green", "EGFP"
)
```

so, i *think* we should basically be able to multiply the integer `F` by the output of `get_emission_events`.  This will currently do this calculation with 1nm bandwidths, since that's how the data comes back from FPbase, but we can elaborate on that depending on whether we want to bin before or after this.

Reply:
misc TODO:
- deal with cases where the fluorophore and excitation spectra are non- or barely-overlapping and fail gracefully.

Reply:
This code will be super useful. Will go through it, and integrate it into the setup with tests.